File: sql-insight-core/src/main/java/org/gongxuanzhang/sql/insight/core/engine/SqlPipeline.java
/*
 * Copyright 2023 java-mysql  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.engine;

import lombok.extern.slf4j.Slf4j;
import org.gongxuanzhang.sql.insight.core.annotation.Temporary;
import org.gongxuanzhang.sql.insight.core.command.Command;
import org.gongxuanzhang.sql.insight.core.engine.json.ExecuteEngine;
import org.gongxuanzhang.sql.insight.core.environment.SqlInsightContext;
import org.gongxuanzhang.sql.insight.core.exception.SqlInsightException;
import org.gongxuanzhang.sql.insight.core.optimizer.Optimizer;
import org.gongxuanzhang.sql.insight.core.optimizer.OptimizerImpl;
import org.gongxuanzhang.sql.insight.core.optimizer.plan.ExecutionPlan;
import org.gongxuanzhang.sql.insight.core.result.ResultInterface;

/**
 * sql lifecycle container
 *
 * @author gongxuanzhangmelt@gmail.com
 **/
@Slf4j
public class SqlPipeline {

    private Optimizer optimizer = new OptimizerImpl();

    private ExecuteEngine executeEngine;

    @Temporary(detail = "how to deal exception?")
    public ResultInterface doSql(String sql) throws SqlInsightException {
        log.info("start execute sql {} ...", sql);
        long startTime = System.currentTimeMillis();

        Command command = optimizer.analysisSql(sql);

        ExecutionPlan plan = optimizer.assign(command);

        ResultInterface resultInterface = executeEngine.executePlan(plan);

        log.info("end sql {} ...take time {}ms", sql.split("\n")[0], (System.currentTimeMillis() - startTime));
        return resultInterface;
    }

    public Optimizer getOptimizer() {
        return optimizer;
    }

    public SqlPipeline setOptimizer(Optimizer optimizer) {
        this.optimizer = optimizer;
        return this;
    }

    public ExecuteEngine getExecuteEngine() {
        return executeEngine;
    }

    public SqlPipeline setExecuteEngine(ExecuteEngine executeEngine) {
        this.executeEngine = executeEngine;
        return this;
    }

    private SqlInsightContext createContext() {
        return null;
    }
}

File: sql-insight-core/src/main/kotlin/tech/insight/core/engine/SqlPipeline.kt
package tech.insight.core.engine

import tech.insight.core.environment.GlobalContext
import tech.insight.core.logging.Logging
import tech.insight.core.logging.TimeReport.timeReport
import tech.insight.core.plan.DMLExecutionPlan
import tech.insight.core.result.ResultInterface
import tech.insight.core.util.truncateStringIfTooLong


/**
 * sql insight core engine
 * execute a sql.
 *
 * sql ---Analyzer--->  command
 * command ---optimizer---->  execute plan
 * execute plan --execute engine---->  result
 *
 * @author gongxuanzhangmelt@gmail.com
 */


/**
 * sql lifecycle container
 */
object SqlPipeline : Logging() {
    init {
        //  static init context
        GlobalContext
    }

    val optimizer: Optimizer = OptimizerImpl
    val analyzer: Analyzer = DruidAnalyzer
    val executeEngine: ExecuteEngine = ExecuteEngineImpl


    fun executeSql(sql: String): ResultInterface {
        val command = timeReport("analysis sql ${truncateStringIfTooLong(sql)}") {
            analyzer.analysisSql(sql)
        }
        val plan = timeReport("optimize command $command") {
            optimizer.optimize(command)
        }
        return timeReport("execute plan $plan") {
            when (plan) {

                is DMLExecutionPlan -> {
                    plan.engine.initSessionContext().use {
                        executeEngine.executePlan(plan)
                    }
                }

                else -> {
                    executeEngine.executePlan(plan)
                }

            }
        }
    }

}





